### PR TITLE
Whitelist constant shape arrays from TODO and character assignment tests

### PR DESCRIPTION
**Pull request by jeanPerier**
- Allow array with constant bounds (can be assumed) to lower
  Keep the rest as a TODO for now.
- Add builder.isComplex for both mlir and fir complex.
- Add character assignement lowering to FIR lit tests.
- Move end-to-end character assignement to lit.

### CODE REVIEW

File: lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -1100,6 +1100,20 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
 
+  /// Temporary helper to detect shapes that do not require evaluating
+  /// bound expressions at runtime or to get the shape from a descriptor.
+  static bool isConstantShape(const Fortran::semantics::ArraySpec &shape) {
+    for (const auto &susbcript : shape) {
+      const auto &lb = susbcript.lbound().GetExplicit();
+      const auto &ub = susbcript.ubound().GetExplicit();
+      if (!lb.has_value() || !Fortran::evaluate::IsConstantExpr(*lb) ||
+          !(susbcript.ubound().isAssumed() ||
+            (ub.has_value() && Fortran::evaluate::IsConstantExpr(*ub))))
```
Comment (schweitzpgi) [inline lib/Lower/Bridge.cpp:1111]: I'd write this as "positive logic" to make it a bit clearer to some future maintainer.  Something like:
```
if (lb && isConst(lb) && (ub && isConst(ub) || isAssumed(ub))) {
  // do nothing
} else {
  return false;
}
```

File: test/Lower/character-assignement.f90 [hunk 2]
```diff
@@ -0,0 +1,103 @@
+! RUN: bbc %s -o "-" -emit-fir | FileCheck %s
```
Comment (schweitzpgi) [inline test/Lower/character-assignement.f90:1]: Probably want to rename this file to remove the `e` between between `assign` and `ment`.

File: test/Lower/end-to-end-character-assignment.f90 [hunk 3]
```diff
@@ -1,3 +1,10 @@
+! RUN: bbc -emit-llvm -o - %s | tco | llc | sed 's/_QP//g' | as -o %t
+! RUN: c++ -std=c++17 %t %S/end-to-end-character-assignment-driver.cpp
```
Comment (schweitzpgi) [inline test/Lower/end-to-end-character-assignment.f90:2]: Nice way around the uniquer. lol

File: lib/Lower/Bridge.cpp (hunk 1 above)
Comment (jeanPerier) [reply lib/Lower/Bridge.cpp]: The logic is not the most readable here indeed... rewritten in a positive way with a lambda helper. Hopefully this logic should disappear once bounds specification expr are supported.

File: test/Lower/character-assignement.f90 (hunk 2 above)
Comment (jeanPerier) [reply test/Lower/character-assignement.f90]: Thanks

File: test/Lower/end-to-end-character-assignment.f90 (hunk 3 above)
Comment (jeanPerier) [reply test/Lower/end-to-end-character-assignment.f90:2]: Yes, that was the laziest way out, the one to mangle for compatibility with pgfortran/gfortran in expression tests is a bit uglier : `sed 's,_QP\([[:alnum:]_]*\),\1_,g'`

**Pull request merged**